import { Controller, Repository } from "@src/core";
import { StepRepository } from "@src/repository";
import { StepServices } from "@src/services";
import { Request, Response } from "express";

export default class StepController extends Controller {

	readonly services: StepServices;
	readonly repository: Repository
	constructor() {
		super();
		this.repository = new StepRepository();
		this.services = new StepServices(this.repository);
	}

	getAll = async (req:Request, res:Response) => {
		try{
			const selectQuery = this.parser(req.query.select as string);
			const steps = await this.services.findAll(selectQuery);
			return this.sendSuccess(res, 200, 'R√©cup√©ration des √©tapes r√©ussis', steps);
		} catch(error) {
			return this.sendError(res, 500, 'Echec lors de la r√©cup√©rations des √©tapes', error);
		}
	}

	getById = async (req:Request, res:Response) => {
		try{
			const { id } = req.params;
			const selectQuery = this.parser(req.query.select as string);
			const step = await this.services.findById(id, selectQuery);
			return this.sendSuccess(res, 200, 'R√©cup√©ration de l\'√©tape r√©ussie', step);
		} catch(error) {
			return this.sendError(res, 500, 'Echec lors de la r√©cup√©ration de l\'√©tape', error);
		} 
	}

	create = async (req:Request, res:Response) => {
		try{
			if(!req.user) return this.sendError(res, 401, 'Vous devez √™tre connect√© pour cr√©er une √©tape', []);
			const data = req.body;
			const authorId = req.user.id;
			const step = await this.services.create(data, authorId);
			console.log("üöÄ ~ file: StepController.ts:43 ~ StepController ~ step:", step)
			return this.sendSuccess(res, 200, 'Cr√©ation de l\'√©tape r√©ussie', step);
		} catch(error) {
			return this.sendError(res, 500, 'Echec lors de la cr√©ation de l\'√©tape', error);
		}
	}

	update = async (req:Request, res:Response) => {
		try {
			const { id } = req.params;
			const data = req.body;
			console.log("üöÄ ~ file: StepController.ts:54 ~ StepController ~ update= ~ data:", data)
			const step = await this.services.update(id, data);
			return this.sendSuccess(res, 200, 'Mise √† jour de l\'√©tape r√©ussie', step);
		} catch (error) {
			return this.sendError(res, 500, 'Echec lors de la mise √† jour de l\'√©tape', error);
		}
	}

	delete = async (req:Request, res:Response) => {
		try {
			const { id } = req.params;
			const step = await this.services.delete(id);
			return this.sendSuccess(res, 200, 'Suppression de l\'√©tape r√©ussie', step);
		} catch (error) {
			return this.sendError(res, 500, 'Echec lors de la suppression de l\'√©tape', error);
		}
	}
}